Discusssion (#1)
andreas-merkel-by/jdatech.github.io: _posts/2020-03-16-dask-usage-at-blue-yonder.markdown

  ---
  layout: single
  title:  "Dask Usage at Blue Yonder"
  date:   2019-10-25 10:00:00 +0100
  tags: technology python data-engineering
  header:
    overlay_image: assets/images/tech_gear_banner.jpg
    overlay_filter: 0.2
    show_overlay_excerpt: false
  author: Andreas Merkel
  author_profile: true
  ---
  # Dask Usage at Blue Yonder

  Back in 2017, Blue Yonder started to look into 
  [Dask/Distributed](https://distributed.dask.org) and how we can leverage it to power
  our machine learning and data pipelines.

>   In 2019, we started to use it heavily in production for providing demand predicions,

Author: typo `predictions`

  recommended order quantities, and price proposals to our customers.

>   Time to take a a look at the way we are using Dask!

Author: I think this could be 
```
It's 2020, and we are using it heavily in production for providing demand predictions, recommending order quantities, and proposing prices to our customers. Let's have a look at how we are doing!
```

  ## Use cases

  First, let's have a look at the use cases we have for Dask. The use of Dask at Blue Yonder
  is strongly coupled to the concept of [Datasets](https://tech.jda.com/introducing-kartothek/),
  tabular data stored in blob storage. We use these datasets for storing everything that is needed
  for generating predictions, recommended order quantities, and recommended prices for our customers,

>   as well as the predictions, order quantities, and prices themselves.

Author: seems redundant to tell about what we do (when we already told in the section `Dask Usage at Blue Yonder`

Reviewer: True, but I'd rather remove it in the first section then.

  ### Downloading data from a database to a dataset

  One typical use case that we have is downloading large amounts of data from a database into a
  dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is

>   an extremely simple: a lot of parallel nodes downloading pieces of data into a blob and one

Author: `and the graph is an extremely simple one: `

OR 

and the graph is simple:

>   reduction node writing the dataset metadata (which is basically a dictionary what data can be found in which blob).

Author: I think it would make sense to show the dask graph (with only 2-4 partitions) to make the end-user "visually" realize how it is a simple graph.

Reviewer: Yes, that's a good idea.

Reviewer: Or maybe even screenshot it from a real Dask instance.

  ### Doing computations on a dataset

>   A lot of our algorithms for machine learning, forecasting, and optimization work on datasets.

Author: seems redundant to tell

Reviewer: Yes, that's true. I was looking for some kind of introductory sentence here.

  A source dataset is read, computations are performed on the data, and the result is written to
  a target dataset.
  In many cases, the layout of the source dataset (the partitioning, i.e., what data resides in which blob)
  is used for parallelization. This means the algorithms work independently on the individual

>   blobs of the source dataset. Therefore our respective Dask graphs again look very simple, with parallel nodes each performing

Author: ```suggestion
  blobs of the source dataset. Therefore, our respective Dask graphs again look very simple, with parallel nodes each performing
```

  the sequential operations of reading in the data from a source dataset blob, doing some computation on it,
  and writing it out to a target dataset blob. Again, there is a final reduction node writing the target
  dataset metadata.

  ### Dataset maintenance

>   We also use Dask/Distributed for performing maintenance operations like garbage collection on datasets.

Author: Given that we are already separating the two points by using headers (which are `Dataset Maintenance` and `Doing computations on a dataset` ), would it make sense to use `we also`?

Reviewer: Was probably from a time before I introduced the headers; removed the also.

  Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
  references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
  In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
  In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.

  ### Repartitioning a dataset

  Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
  the case that we have the data available in a suitable partitioning, for instance, if the data results
  from a computation that used a different partitioning. In this case, we have to repartition the data.

>   This works by reading the dataset as a Dask distributed dataframe (ddf), repartitioning the ddf using

Author: ```suggestion
  This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) repartitioning the dataframe using
```

  network shuffle, and writing it out again to a dataset.

  ## Dask cluster setup at Blue Yonder

  We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
  of our domain and partly historically grown. We are running a large number of individual clusters
  that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
  setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number

>   of upstream pull requests.

Author: I think mentioning stuff like Semaphore implemented by us on distributed would be very beneficial. Would be helpful to make it analogous to a similar feature in PySpark (if it has)

Reviewer: I agree that it's a good idea to include references to PRs we've done, but I think the semaphore is not an example of an edge case we have run into because of our cluster setup, but rather a missing feature we needed. But maybe we could add a separate paragraph on that?

  ### Cluster separation

  We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
  For the same reason, we also run separate clusters for production, staging, and development environments. 
  But it does not stop there. For performance reasons, we install the Python packages holding the code needed
  for the computations on each worker. Since the different components of our compute pipeline are implemented
  and maintained as completely separate software packages whose requirements are not synchronized, this also

>   means we have to run a separate Dask cluster for each of the components. This leads to us operating more than

Author: I don't think it is easy to understand why we need to have different clusters are component level also.

Author: typo *at

Reviewer: Yes, it's a bit tricky. Maybe an example could help here?

  ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
  and computing something. While this leads to overhead in terms of administration and hardware resouces,
  it also gives us a lot of flexibility. For instance, we can update the software on the cluster of one part of the compute pipeline

>   while another part of the pipeline is computing something on a different cluster.

Author: I think cluster separation is something only experienced folks can really understand the importance/complexity off.

In order to make the reasoning more obvious, we start off by first telling how having a handful of dedicated clusters would not work out. Then we proceed by telling how we solve this by having isolation not only at the customer level, but also environment level and component level. And then we proceed with how the advantages outweighs the costs of doing like this.

Reviewer: sounds good 👍

>   ### Some numbers

Author: Do they come under `use cases`?

Reviewer: I put them there because we take the numbers up again in the following sections, but I also see the point that we could move the numbers more to the beginning.

  The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
  of the computation being performed, its resource requirements, and the available timeframe for the computation.
  At the time of writing this, we are running more than 500 distinct clusters.
  They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.

  ## Cluster scaling and resilience

  To improve manageability, resilience, and resource utilization, we run the Dask clusters on top
  of [Apache Mesos](http://mesos.apache.org/)/[Aurora](http://aurora.apache.org/). This

>   means it is really easy to commission or decommission Dask clusters and to add or remove workers to/from

Author: I don't know about Mesos/Aurora enough, but will it be possible to explain with example/image?

Author: actually, just saw line 101-103. Could we move this to this para?

Reviewer: Actually, it refers to deployment and autoscaling, I'll move it to the respective places.

  an existing Dask cluster. At the moment, we are starting a migration from Mesos/Aurora towards 

>   [Kubernetes](https://kubernetes.io/)

Author: given that the article focus is on "usage of dask at Blue Yonder", my opinion would be to remove the last line.

Reviewer: Yes, it's pretty unrelated. I somehow want to mention Kubernetes, as it's the Future. I'll just add it to the list of cluster managers in the sentence before.

  Running on top of a system like Mesos or Kubernetes provides us with resilience since a failing worker 
  (for instance, as result of a failing hardware node)
  can simply be restarted on another node of the system. 

  Running 500 Dask clusters requires a lot of hardware. We have put two measures in place to improve
  the utilization of hardware resources: oversubscription and autoscaling.

  ### Oversubscription

  [Oversubscription](http://mesos.apache.org/documentation/latest/oversubscription/) is a feature of Mesos

>   that allows to allocate more resources than physically present to services running on the system.

Author: ```suggestion
  that allows allocating more resources than physically present to services running on the system.
```

  This is based on the assumption that not all services exhaust all of their allocated resources at the same time.

>   If the assumption is violated, classifying services into tiers allows to preempt less important services

Author: ```suggestion
  If the assumption is violated, we prioritize the resouces to more important ones
```

Author: typo in `resources` in my comment

  in favor of critical ones.
  We use this to re-purpose the resources allocated for production clusters but not utilized the whole time
  and use them for development and staging systems.

  ### Autoscaling

>   Autoscaling is a mechanism we implemented to dynamically adapt the number of workers in a Dask cluster

Author: Do we work on an assumption people know what autoscaling is? I don't know if it is common lingo in the data-engg world.

Reviewer: The first sentence is intended as a definition of the term, as it is not common, but was created by us.

  to the load on the cluster. For this, we added the ``desired_workers`` metric to Distributed, which exposes
  the degree of parallelism that a computation has and allows us to infer how much workers a cluster should
  ideally have. Based on this metric, as well as on the overall resources available and on fairness criteria

>   (remember, we run a lot of Dask/Distributed clusters), we add or remove workers to our clusters.

Author: Is there any formulae/logic that we work with. Then telling that (if we are allowed) would be pretty interesting, I think

Reviewer: There even is a paper describing the logic we implement. I'll dig it out...
